Event and alarm management

bhaveshdell opened the pull request:
Why I did it
This PR contains code changes for providing extension to the Event Framework as specified in the https://github.com/sonic-net/SONiC/pull/1409

How I did it
Followed design specified in https://github.com/sonic-net/SONiC/pull/1409.

How to verify it
Unit tests added.

This PR was originally reviewed and merged.
https://github.com/sonic-net/sonic-buildimage/pull/17949

However, due to build error post-merge the PR was reverted.
https://github.com/sonic-net/sonic-buildimage/pull/20052

The possible fix for build error has been fixed as per comments in PR
https://github.com/sonic-net/sonic-buildimage/pull/20064

bhaveshdell committed:
Support for event persistence in redis-db.

bhaveshdell committed:
Updates

bhaveshdell committed:
Updates including fix to  eventdb in test enviroment.

bhaveshdell committed:
Add sonic yang to model event and alarm table.
remove ack, noack from sonic-common-event yang.

bhaveshdell committed:
Add event/alarm  persistence related testscases

bhaveshdell committed:
Remove file eventdb_ut.cpp.

bhaveshdell committed:
Updates to eventdb testsuite.

bhaveshdell committed:
Revert changes to existing eventd UT.
Set eventdb testcases as separate test binary.
Skip testcase execution if DB connection failure.

bhaveshdell committed:
Commit test related config files.

bhaveshdell committed:
Local make customizations.

bhaveshdell committed:
wRevert "Local make customizations."

This reverts commit e9425a0aad5d9413efaaf1b4eccd6236aef50aba.

bhaveshdell committed:
Separate eventd and eventdb targets in the makefile.
Set eventdb process startretries= 0 and expected exit code=0.
This is since if no event types are defined in the event_profile, the process can exit.

mssonicbld commented:
/azp run Azure.sonic-buildimage

bhaveshdell commented:
/azp run Azure.sonic-buildimage

bhaveshdell commented:
/AzurePipelines run Azure.sonic-buildimage

prsunny commented:
/azp run Azure.sonic-buildimage

prsunny commented:
> /AzurePipelines run Azure.sonic-buildimage

@bhaveshdell , you can try with `/azpw run Azure.sonic-buildimage`

venkatmahalingam commented:
/azpw run Azure.sonic-buildimage

mssonicbld commented:
Only PR owner can use /azpw run

bhaveshdell commented:
/azpw run Azure.sonic-buildimage

mssonicbld commented:
/AzurePipelines run Azure.sonic-buildimage

zhangyanzhao commented:
/azpw run Azure.sonic-buildimage

mssonicbld commented:
Only PR owner can use /azpw run

qiluo-msft commented:
/azp run Azure.sonic-buildimage

qiluo-msft commented on src/sonic-eventd/src/loghandler.h:
```diff
@@ -0,0 +1,5 @@
+#include <iostream>
+extern "C" void openSyslog();
+extern "C" void writeToSyslog(std::string ev_id, int ev_sev, std::string ev_type, std::string ev_act, std::string ev_msg, std::string ev_static_msg);
```
this is valid comment.

qiluo-msft commented on src/sonic-eventd/src/eventutils.h:
```diff
@@ -0,0 +1,43 @@
+#include <string>
+#include <unordered_map>
+
+using namespace std;
+
+const string EVENT_SEVERITY_CRITICAL_STR = "CRITICAL";
+const string EVENT_SEVERITY_MAJOR_STR    = "MAJOR";
+const string EVENT_SEVERITY_MINOR_STR    = "MINOR";
+const string EVENT_SEVERITY_WARNING_STR  = "WARNING";
+const string EVENT_SEVERITY_INFORMATIONAL_STR = "INFORMATIONAL";
+
+const string EVENT_ENABLE_TRUE_STR  = "true";
+const string EVENT_ENABLE_FALSE_STR = "false";
+
+const string EVENT_ACTION_RAISE_STR  = "RAISE";
+const string EVENT_ACTION_CLEAR_STR  = "CLEAR";
+const string EVENT_ACTION_ACK_STR    = "ACKNOWLEDGE";
+const string EVENT_ACTION_UNACK_STR  = "UNACKNOWLEDGE";
+
```
this is a valid comment

venkatmahalingam replied:
Will address the comments

bhaveshdell committed:
Address review comments.
Avoid using namespace std; in a header file to prevent namespace pollution; qualify standard types explicitly.
Avoid non-async-signal-safe function inside signal handler.

bhaveshdell committed:
Updates.

bhaveshdell committed:
Remove debug statements.

bhaveshdell committed:
Add script to start eventdb.
Script starts eventdb only if atleast one event is defined in the profile.
Add validation around json operations.

mssonicbld commented:
/azp run Azure.sonic-buildimage

qiluo-msft commented on src/sonic-eventd/src/eventutils.cpp:
```diff
@@ -0,0 +1,103 @@
+    try {
+      j = json::parse(ifs);
+    }
+    catch (const std::exception &e) {
```
>exception [](http://example.com/codeflow?start=22&length=9)

Could you use more specific exception class?

bhaveshdell replied:
Updated to handle specific exception

bhaveshdell committed:
Handle specific exception while parsing json file.
Fix indentation.

bhaveshdell committed:
Add the new yangs to setup.py

mssonicbld commented:
/azp run Azure.sonic-buildimage

bhaveshdell committed:
Add the sonic-alarm.yang and sonic-event.yang as NON_CONF_YANG_FILES

bhaveshdell commented on src/sonic-eventd/src/eventutils.cpp:
```diff
@@ -0,0 +1,103 @@
+
+    json j;
+    try {
+      j = json::parse(ifs);
```
Code updated.

zbud-msft commented on src/sonic-eventd/src/loghandler.cpp:
```diff
@@ -0,0 +1,48 @@
+    openlog (NULL, LOG_CONS | LOG_PID | LOG_NDELAY, LOG_LOCAL4);
+}
+
+extern "C" void writeToSyslog(char* ev_id, int ev_sev, char* ev_type, char* ev_act,
```
.h definition takes in const char*, shouldnt this follow?

bhaveshdell replied:
Fixed.

bhaveshdell committed:
Address review comments.
Fix event purge time computation.
Fix stats update on purge.

bhaveshdell commented:
/azpw run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

bhaveshdell commented on dockers/docker-database/database_config.json.j2:
```diff
@@ -147,6 +147,12 @@
         }
 {% endif %}
+        ,
+        "EVENT_DB" : {
```
This daemon listens to eventd and writes to DB. 
In case eventd is shutdown, no events are logged in the DB.

qiluo-msft replied:
what happens when INCLUDE_EVENTD=n, does the EVENT_DB instance still get provisioned? Does eventdb_wrapper.sh still run?

Could you guard your new section by `{% if INCLUDE_EVENTD == "y" %}`?

bhaveshdell replied:
Fixed

mssonicbld commented:
/azp run Azure.sonic-buildimage

yxieca commented:
/azpw ms_conflict

bhaveshdell committed:
Resolve build errors.

mssonicbld commented:
/azp run Azure.sonic-buildimage

qiluo-msft commented on src/sonic-eventd/src/eventconsume.h:
```diff
@@ -0,0 +1,56 @@
+#include <swss/dbconnector.h>
+#include <swss/subscriberstatetable.h>
+
+using namespace swss;
```
namespace pollution. Please remove `using namespace` in all header files

bhaveshdell replied:
Fixed

qiluo-msft commented on files/build_templates/docker_image_ctl.j2:
```diff
@@ -110,10 +110,18 @@ function preStartAction()
             if [[ "$DATABASE_TYPE" != "bmcdb" ]]; then
                 docker cp /tmp/dump.rdb database$DEV:/var/lib/redis_bmp/
             fi
+            COLD_DIR=/host/coldboot
```
changes to this file is not related to your motivation. why changing?

bhaveshdell replied:
This is part of the EVENT_DB persistence support. But persistence is not targeted for this commit. Will revert.

qiluo-msft commented on src/sonic-eventd/src/eventdb.cpp:
```diff
@@ -0,0 +1,29 @@
+
+int main()
+{
+    swss::Logger::getInstance().setMinPrio(swss::Logger::SWSS_DEBUG);
```
SWSS_DEBUG will make it too noisy. Normally just use SWSS_NOTICE

qiluo-msft commented on src/sonic-eventd/src/loghandler.cpp:
```diff
@@ -0,0 +1,48 @@
+        syslog(LOG_MAKEPRI(ev_sev, SYSLOG_FACILITY), LOG_FORMAT,
+            safe(ev_type), safe(ev_id), safe(ev_src), safe(ev_static_msg), safe(ev_msg));
+    } else {
+        const char LOG_FORMAT[] = "[%s] (%s), %%%s: %%%%%s: %s %s"; 
```
The %% is applied for help with parsing -> this seems an adhoc design, and fragile.
Why not reuse the serialization/deserialization in sonic-eventd? @zbud-msft please help review here.

bhaveshdell committed:
fix: Fix review comments

- Remove namespace pollution from eventconsume.h and fix build issues
- Add explicit swss:: and std:: prefixes to maintain type safety
- Optimize clearAckAlarmStatistic with early exit flag

bhaveshdell committed:
Add timeout to event receive blocking call.
Updates to the eventdb testsuite.

bhaveshdell committed:
Remove loading of eventdb on database start.

bhaveshdell committed:
Guard creation of EVENTDB with flag include_system_eventdb.

venkatmahalingam commented:
@qiluo-msft  @make1980 @zbud-msft Can you take a look at the new changes, if no more comments, please approve and merge PR.

mssonicbld commented:
⚠️ **Notice**: `/azpw run` only runs failed jobs now. If you want to trigger a whole pipline run, please rebase your branch or close and reopen the PR.
💡 **Tip**: You can also use `/azpw retry` to retry failed jobs directly.

Retrying failed(or canceled) jobs...

bhaveshdell commented on src/sonic-eventd/src/loghandler.cpp:
```diff
@@ -0,0 +1,48 @@
+        syslog(LOG_MAKEPRI(ev_sev, SYSLOG_FACILITY), LOG_FORMAT,
+            safe(ev_type), safe(ev_id), safe(ev_src), safe(ev_static_msg), safe(ev_msg));
+    } else {
+        const char LOG_FORMAT[] = "[%s] (%s), %%%s: %%%%%s: %s %s"; 
```
@qiluo-msft JSON serialization would be a better option for parsing, and we'll consider it in the next PR.

mssonicbld commented:
Retrying failed(or canceled) stages in build 1097133:


✅Stage **Test**:
- Job impacted-area-kvmtest-t2 by Elastictest: retried.
- Job kvmtest-t1-lag-vpp by Elastictest: retried.
- Job impacted-area-kvmtest-dpu by Elastictest: retried.
- Job impacted-area-kvmtest-multi-asic-t1 by Elastictest: retried.
- Job impacted-area-kvmtest-t0 by Elastictest: retried.
- Job impacted-area-kvmtest-t0-2vlans by Elastictest: retried.
- Job impacted-area-kvmtest-t1-lag by Elastictest: retried.
- Job impacted-area-kvmtest-t0-sonic by Elastictest: retried.

mssonicbld commented:
Retrying failed(or canceled) jobs...

mssonicbld commented:
Retrying failed(or canceled) stages in build 1097133:


✅Stage **Test**:
- Job impacted-area-kvmtest-t0-2vlans by Elastictest: retried.
- Job impacted-area-kvmtest-t0-sonic by Elastictest: retried.
- Job impacted-area-kvmtest-t1-lag by Elastictest: retried.
- Job impacted-area-kvmtest-t2 by Elastictest: retried.
- Job impacted-area-kvmtest-dpu by Elastictest: retried.
- Job impacted-area-kvmtest-t0 by Elastictest: retried.
- Job kvmtest-t1-lag-vpp by Elastictest: retried.
- Job impacted-area-kvmtest-multi-asic-t1 by Elastictest: retried.

venkatmahalingam commented:
/azpw retry

mssonicbld commented:
Retrying failed(or canceled) jobs...

bhaveshdell committed:
Pass include_system_eventd to runtime jinjanate rendering

- database_config.json.j2: Add 'is defined' guard for include_system_eventd
- Dockerfile.j2: Add ENV INCLUDE_SYSTEM_EVENTD to bake build-time flag into container
- docker-database-init.sh: Pass include_system_eventd to jinjanate calls

mssonicbld commented:
Retrying failed(or canceled) stages in build 1111176:


✅Stage **Test**:
- Job Trivy vulnerability scan (docker-ptf): skipped (optional job and does not block the PR merge).
- Job impacted-area-kvmtest-t0 by Elastictest: retried.
